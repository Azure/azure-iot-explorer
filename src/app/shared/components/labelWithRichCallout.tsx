/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Callout, ILabelProps, Label, IconButton, DirectionalHint, getId } from '@fluentui/react';
import { INFO } from '../../constants/iconNames';
import '../../css/_labelWithTooltip.scss';

export interface LabelWithRichCalloutProps extends ILabelProps {
    calloutContent: JSX.Element;
    directionalHint?: DirectionalHint;
}

export interface LabelWithRichCalloutState{
    showCallout: boolean;
}

export const LabelWithRichCallout: React.FC<LabelWithRichCalloutProps> = React.memo((props: LabelWithRichCalloutProps) => {
    const { calloutContent, directionalHint, style } = props;
    const [ showCallout, setShowCallout ] = React.useState<boolean>(false);

    const setVisibility = () => setShowCallout(true);
    const dismissCallout = () => setShowCallout(false);
    const buttonId = getId('iconbutton');

    return (
        <div
            style={style}
            className="labelWithTooltip"
        >
            <Label
                {...props}
            />
            {calloutContent &&
                <>
                    <IconButton
                        iconProps={{ iconName: INFO }}
                        id={buttonId}
                        onClick={setVisibility}
                    />
                    { showCallout && <Callout
                        role="alertdialog"
                        gapSpace={0}
                        target={`#${buttonId}`}
                        setInitialFocus={true}
                        directionalHint={directionalHint || DirectionalHint.rightCenter}
                        hidden={!showCallout}
                        onDismiss={dismissCallout}
                    >
                        {calloutContent}
                    </Callout>}
                </>
            }
        </div>
    );
});
