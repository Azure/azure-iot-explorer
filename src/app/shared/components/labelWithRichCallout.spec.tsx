/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { Callout, IconButton } from '@fluentui/react';
import { LabelWithRichCallout } from './labelWithRichCallout';

describe('components/shared/labelWithRichCallout', () => {

    it('matches snapshot', () => {
        const wrapper = shallow(
            <LabelWithRichCallout
                calloutContent={<></>}
            >
                {'labelText'}
            </LabelWithRichCallout>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('calls expected functions', () => {
        const wrapper = mount(
            <LabelWithRichCallout
                calloutContent={<></>}
            >
                {'labelText'}
            </LabelWithRichCallout>
        );

        act(() => wrapper.find(IconButton).props().onClick(undefined));
        wrapper.update();
        const callout = wrapper.find(Callout);
        expect(callout).toBeDefined();

        act((() => callout.props().onDismiss(null));
        wrapper.update();
        const updatedCallout = wrapper.find(Callout);
        expect(updatedCallout.length).toEqual(0);
    });
});
