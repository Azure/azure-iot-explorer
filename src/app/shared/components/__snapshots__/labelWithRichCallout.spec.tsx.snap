// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/shared/labelWithRichCallout matches snapshot 1`] = `
<div
  className="labelWithTooltip"
>
  <StyledLabelBase
    calloutContent={<React.Fragment />}
  >
    labelText
  </StyledLabelBase>
  <CustomizedIconButton
    iconProps={
      Object {
        "iconName": "Info",
      }
    }
    id="iconbutton0"
    onClick={[Function]}
  />
</div>
`;
